content/4_Gold/Ternary_Search.mdx

```diff
@@ -0,0 +1,142 @@
+---
+id: ternary-search
+title: Ternary Search
+author: Chongtian Ma
+prerequisites:
+  - /silver/binary-search
+description: Searching for a peak in a unimodal function.
+frequency: 1
+---
+
+<Resources>
+	<Resource
+		source="CP-Algorithms"
+		title="Ternary Search"
+		url="https://cp-algorithms.com/num_methods/ternary_search.html"
+		starred
+	/>
+</Resources>
+
+Ternary search is the equivalent of binary searching for the zero on the
+derivative function. The most basic function can be represented by the standard
+form of a parabola: $ax^2+bx+c$. Depending on if $a$ is positive or negative,
+you are searching for the local minimum or maximum of the function,
+respectively. Note that the function must be unimodal (i.e. change directions in exactly one place)!
+
+Unlike binary search where a checking function $f(x)$ only needs to return true
+or false, we need $f(x)$ to output an integer. Between the bounds of $[l, r]$,
+we need to compare the integer output of $f(m_1)$ with $f(m_2)$ where
+$l < m_1 < m_2 < r$ to determine how $l$ and $r$ should be moved.
+
+For example, if the function has a local maximum between $[l, r]$ and
+$f(m_1) < f(m_2)$, then we know the local maximum cannot be between $[l, m_1]$,
+so we continue searching in $[m_1, r]$. For a mini-proof, consider the following
+cases:
+
+- Let's denote the local maximum as $g$. If $m_1$ and $m_2$ are less than $g$,
+  this means that $f(m_1)$ and $f(m_2)$ are both on the increasing side of the
```

Comment:
you glossed over the possibility of the function being flat on an interval, but this is fine as a rough sketch

Reply:
it shouldn't affect things and should be quite intuitive, no? maybe leave it as an exercise to the reader? 😄

Reply:
then you should explicitly state that you are leaving out this possibility